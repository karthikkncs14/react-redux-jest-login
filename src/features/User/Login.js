import React, {Fragment, useEffect} from 'react';
import {Link} from 'react-router-dom';
import {useForm} from 'react-hook-form';
import {useSelector, useDispatch} from 'react-redux';
import {loginUser, userSelector, clearState} from '../../Api/UserSlice';
import {useHistory} from 'react-router-dom';
import Input from '../../Components/Input';
import Button from '../../Components/Button';
import User from '../../assets/user.svg';
import Form from "../../Components/Form";
import Header from "../Header";

const Login = ({}) => {
    const dispatch = useDispatch();
    const history = useHistory();
    // const {register, errors, handleSubmit} = useForm();
    const {register, handleSubmit, formState: {errors}, reset} = useForm();
    const {isFetching, isSuccess, isError, errorMessage} = useSelector(
        userSelector
    );
    //console.log("isError",isError,isSuccess);
    const onSubmit = async (data) => {
        try {
            await dispatch(loginUser(data));
        } catch (e) {

        }
    };

    useEffect(() => {
        return () => {
            dispatch(clearState());
        };
    }, []);

    useEffect(() => {
        if (isSuccess) {
            dispatch(clearState());
            history.push('/');
        }
    }, [isSuccess]);

    return (
        <div className="main-container">
            <Header><div className="form-header">Login</div></Header>

            <div className={"form-container"}>
                {/*<div className="col-sm-4">col-sm-4</div>*/}

                <Form onSubmit={handleSubmit(onSubmit)}>
                    <img src={User} alt="User" style={{"paddingBottom":"33px"}}/>
                    <div className="error-alert">{isError && errorMessage}</div>
                    <div className={"form-col"}>
                        <label htmlFor="email">Email address</label>
                        <Input disabled={isFetching} placeholder={"Email address..."} id="email"
                               name="email"
                               register={register}
                               type="email"/>
                        {errors.email && <span className="error-alert">{"Please enter email"}</span>}
                    </div>
                    <div className={"form-col"}>
                        <label htmlFor="password">Password</label>
                        <Input disabled={isFetching}  placeholder={"Password..."}
                               id="password"
                               name="password"
                               register={register}
                               type="password"
                        />
                        {errors.password && <span className="error-alert">{"Please enter password"}</span>}
                    </div>
                    <div className="form-submit">
                        <Button disabled={isFetching} type="submit">Login</Button>
                    </div>
                    <div>
                        {!isFetching ?
                        <Link to="SignUp">No account? registered</Link> :
                        <span style={{color: "#c6c6c6"}}>No account? registered</span>}
                    </div>

                </Form>
            </div>
        </div>
    );
};

export default Login;
