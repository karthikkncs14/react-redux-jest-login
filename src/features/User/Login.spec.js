import React from "react";
import {shallow, mount} from "enzyme";
import configureStore from "redux-mock-store";

import * as ReactReduxHooks from "react-redux";
import Login from "./Login";
import {configure} from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import userReducer, {loginUser, userSelector, clearState} from './UserSlice';
import {BrowserRouter as Router, Switch, Route, Link} from 'react-router-dom';
import {render, screen, fireEvent, act} from "@testing-library/react";

configure({adapter: new Adapter()});

describe("Login", () => {
    let wrapper;
    let useEffect;
    let store;

    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    };
    beforeEach(() => {
        store = configureStore()({
            username: '',
            email: '',
            isFetching: false,
            isSuccess: false,
            isError: false,
            errorMessage: '',
        });

        useEffect = jest.spyOn(React, "useEffect");
        mockUseEffect();
        mockUseEffect();

        jest
            .spyOn(ReactReduxHooks, "useSelector")
            .mockImplementation(state => store.getState());

        jest
            .spyOn(ReactReduxHooks, "useDispatch")
            .mockImplementation(() => store.dispatch);

        wrapper = shallow(<Login store={store}/>);
        render(<Router><Login/></Router>);
    });

    describe("on start", () => {
        it("should render button", () => {
            expect(wrapper.find("button")).toHaveLength(1);
        });

        it("should not trigger submit", async () => {
            fireEvent.input(screen.getByRole("textbox", {name: /email/i}), {
                target: {
                    value: "test"
                }
            });

            fireEvent.input(screen.getByLabelText("password"), {
                target: {
                    value: "password"
                }
            });
            await act( async () => {
                fireEvent.submit(screen.getByRole("button"));
            });
            expect(screen.getByRole("textbox", {name: /email/i}).value).toBe("test");
            expect(screen.getByLabelText("password").value).toBe("password");
        });

        it("should check proper input value", async () => {
            fireEvent.input(screen.getByRole("textbox", {name: /email/i}), {
                target: {
                    value: "karthikkncs14@gmail.com"
                }
            });

            fireEvent.input(screen.getByLabelText("password"), {
                target: {
                    value: "123456"
                }
            });


            await act( async () => {
                fireEvent.submit(screen.getByRole("button"));
            });
            expect(screen.getByRole("textbox", {name: /email/i}).value).toBe("karthikkncs14@gmail.com");
            expect(screen.getByLabelText("password").value).toBe("123456");
        });

        it("should trigger submit button", async () => {
            fireEvent.input(screen.getByRole("textbox", {name: /email/i}), {
                target: {
                    value: "karthikkncs14@gmail.com"
                }
            });

            fireEvent.input(screen.getByLabelText("password"), {
                target: {
                    value: "123457"
                }
            });
            await act( async () => {
                fireEvent.submit(screen.getByRole("button"));
            });
        });
    });
});
