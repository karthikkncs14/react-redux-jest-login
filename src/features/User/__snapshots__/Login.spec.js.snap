// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWorld renders user with skills 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-50 flex flex-col justify-center py-12 sm:px-6 lg:px-8"
  >
    <div
      class="sm:mx-auto sm:w-full sm:max-w-md"
    >
      <h2
        class="mt-6 text-center text-3xl font-extrabold text-gray-900"
      >
        Sign in to your account
      </h2>
    </div>
    <div
      class="mt-8 sm:mx-auto sm:w-full sm:max-w-md"
    >
      <div
        class="bg-white py-8 px-4 shadow sm:rounded-lg sm:px-10"
      >
        <form
          class="space-y-6"
          method="POST"
        >
          <div>
            <label
              class="block text-sm font-medium text-gray-700"
              for="email"
            >
              Email address
            </label>
            <div
              class="mt-1"
            >
              <input
                autocomplete="email"
                class="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
                id="email"
                name="email"
                required=""
                type="email"
              />
            </div>
          </div>
          <div>
            <label
              class="block text-sm font-medium text-gray-700"
              for="password"
            >
              Password
            </label>
            <div
              class="mt-1"
            >
              <input
                autocomplete="current-password"
                class="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
                id="password"
                name="password"
                required=""
                type="password"
              />
            </div>
          </div>
          <div>
            <button
              class="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
              type="submit"
            >
              Sign in
            </button>
          </div>
        </form>
        <div
          class="mt-6"
        >
          <div
            class="relative"
          >
            <div
              class="relative flex justify-center text-sm"
            >
              <span
                class="px-2 bg-white text-gray-500"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HelloWorld renders user without skills 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-50 flex flex-col justify-center py-12 sm:px-6 lg:px-8"
  >
    <div
      class="sm:mx-auto sm:w-full sm:max-w-md"
    >
      <h2
        class="mt-6 text-center text-3xl font-extrabold text-gray-900"
      >
        Sign in to your account
      </h2>
    </div>
    <div
      class="mt-8 sm:mx-auto sm:w-full sm:max-w-md"
    >
      <div
        class="bg-white py-8 px-4 shadow sm:rounded-lg sm:px-10"
      >
        <form
          class="space-y-6"
          method="POST"
        >
          <div>
            <label
              class="block text-sm font-medium text-gray-700"
              for="email"
            >
              Email address
            </label>
            <div
              class="mt-1"
            >
              <input
                autocomplete="email"
                class="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
                id="email"
                name="email"
                required=""
                type="email"
              />
            </div>
          </div>
          <div>
            <label
              class="block text-sm font-medium text-gray-700"
              for="password"
            >
              Password
            </label>
            <div
              class="mt-1"
            >
              <input
                autocomplete="current-password"
                class="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
                id="password"
                name="password"
                required=""
                type="password"
              />
            </div>
          </div>
          <div>
            <button
              class="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
              type="submit"
            >
              Sign in
            </button>
          </div>
        </form>
        <div
          class="mt-6"
        >
          <div
            class="relative"
          >
            <div
              class="relative flex justify-center text-sm"
            >
              <span
                class="px-2 bg-white text-gray-500"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`isUserConnected: true should render a disconnect button 1`] = `
"<ContextProvider value={{...}}>
  <Login />
</ContextProvider>"
`;
